aptos-move/framework/aptos-framework/sources/HashMap.move

```diff
@@ -0,0 +1,291 @@
+    /// Note it may not split the actual overflowed bucket.
+    /// Abort if `key` already exists.
+    public fun insert<K, V>(map: &mut HashMap<K, V>, key: K, value: V) {
+        let hash = sip_hash(&key);
+        let index = bucket_index(map.level, map.num_buckets, hash);
+        let bucket = Table::borrow_mut(&mut map.buckets, &index);
+        let i = 0;
+        let len = Vector::length(bucket);
+        while (i < len) {
+            let entry = Vector::borrow(bucket, i);
+            assert!(&entry.key != &key, 0);
+            i = i + 1;
+        };
+        Vector::push_back(bucket, Entry {hash, key, value});
+        map.len = map.len + 1;
+
+        if (load_factor(map) > SPLIT_THRESHOLD) {
+            split_one_bucket(map);
+        }
+    }
+
+    /// Split the next bucket into two and re-insert existing items.
+    fun split_one_bucket<K, V>(map: &mut HashMap<K, V>) {
+        let new_bucket_index = map.num_buckets;
+        // the next bucket to split is num_bucket without the most significant bit.
+        let to_split = new_bucket_index ^ (1 << map.level);
+        let new_bucket = Vector::empty();
+        map.num_buckets = new_bucket_index + 1;
+        // if the whole level is splitted once, bump the level.
+        if (to_split + 1 == 1 << map.level) {
+            map.level = map.level + 1;
+        };
+        let old_bucket = Table::borrow_mut(&mut map.buckets, &to_split);
+        // partition the bucket, so the [0..i) stays in old bucket, [j..len] goes to new bucket
+        let i = 0;
+        let j = Vector::length(old_bucket);
+        let len = j;
+        while (i < j) {
+            let entry = Vector::borrow(old_bucket, i);
+            let index = bucket_index(map.level, map.num_buckets, entry.hash);
+            if (index == new_bucket_index) {
+                j = j - 1;
+                Vector::swap(old_bucket, i, j);
+            } else {
+                i = i + 1;
+            };
+        };
+        while (j < len) {
+            let entry = Vector::pop_back(old_bucket);
+            Vector::push_back(&mut new_bucket, entry);
+            len = len - 1;
+        };
+        Table::add(&mut map.buckets, &new_bucket_index, new_bucket);
+    }
+
+    /// Return the expected bucket index to find the hash.
+    fun bucket_index(level: u8, num_buckets: u64, hash: u64): u64 {
+        let index = hash % (1 << (level + 1));
+        if (index < num_buckets) {
+            // in existing bucket
+            index
+        } else {
+            // in unsplitted bucket
+            index % (1 << level)
+        }
+    }
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    /// The requirement of &mut HashMap is to bypass the borrow checker issue described in https://github.com/move-language/move/issues/95
+    public fun borrow<K, V>(map: &mut HashMap<K, V>, key: &K): &V {
```

Comment:
Don't use &K, it will run in the same problem with the borrow checker we already discussed.

Reply:
it works if we use &mut (see https://github.com/move-language/move/issues/95#issuecomment-1123015047), the downside is that we can only borrow once from the struct, but I haven't had use case that requires borrow multiple items in the same time

Reply:
From my Rust experience, `&mut` is a *very* different thing than `&`. Making this `&mut` when it doesn't need to be is also very confusing to users. Also we already decided to make Tables take `K` and this flavor of table should align. We can introduce refs everywhere once we solved the issue.

Reply:
sounds goods to me

Reply:
Note that the bucket_index is still a local var, so we still need &mut until Table supports borrow by K

Reply:
Got it. Will someone at Aptos go and update the Table extension in the way we discussed last week?

Reply:
as soon as https://github.com/move-language/move/pull/109 is merged, we can upgrade aptos to use latest revision